common/queue.h

```diff
@@ -86,6 +87,72 @@ class Queue {
 	List<T>	_impl;
 };
 
+/**
+ * Queue ordered by a provided priority function
+ * NOTE: Unlike in the C std library, we have to provide a comparator that sorts
+ * the array, so that the smallest priority comes last
+ */
+template <class _Ty, class _Container = Common::Array<_Ty>, class _Pr = Common::Less<_Ty>>
+class PriorityQueue {
+public:
+	typedef const _Ty& const_reference;
+
+public:
+	PriorityQueue() : c(), comp() {}
+
+	explicit PriorityQueue(const _Pr &_Pred) : c(), comp(_Pred) {}
+
+	PriorityQueue(const _Pr &_Pred, const _Container &_Cont) : c(_Cont), comp(_Pred) {
+		make_heap(c.begin(), c.end(), comp);
+	}
+
+	template <class _InIt>
+	PriorityQueue(_InIt _First, _InIt _Last, const _Pr &_Pred, const _Container &_Cont) : c(_Cont), comp(_Pred) {
+		c.insert(c.end(), _First, _Last);
+		make_heap(c.begin(), c.end(), comp);
+	}
+
+	template <class _InIt>
+	PriorityQueue(_InIt _First, _InIt _Last) : c(_First, _Last), comp() {
+		make_heap(c.begin(), c.end(), comp);
+	}
+
+	template <class _InIt>
+	PriorityQueue(_InIt _First, _InIt _Last, const _Pr &_Pred) : c(_First, _Last), comp(_Pred) {
+		make_heap(c.begin(), c.end(), comp);
+	}
+
+	bool empty() const {
+		return c.empty();
+	}
+
+	size_t size() const {
+		return c.size();
+	}
+
+	const_reference top() const {
+		return c.back();
+	}
+
+	void push(const typename _Container::value_type &_Val) {
+		c.push_back(_Val);
+		Common::sort(c.begin(), c.end(), comp);
```

Comment:
Using a sorted array is a rather inefficient way to implement a priority queue. E.g. `push` is $O(n)$ (or worse, if you `sort` is not clever enough to deal with an almost sorted list optimally) instead of $O(1)$ as it would be for e.g. a heap based priority queue. I see that at least one removed implementation uses a heap; why not use that as reference?